### PR TITLE
Add multi-column pagination support

### PR DESCRIPTION
**Pull request by smklein**
See https://github.com/oxidecomputer/omicron/pull/512 for context / desired usage.

This extends the `paginated` function to a two-column variant, allowing sorting through multiple columns simultaneously.

### CODE REVIEW

**Review by davepacheco — COMMENTED** [review]
Nice!  I'm curious how you figured out what the appropriate trait bounds are.

Is it worth adding a test?  I'm not sure if it's worth generating a whole fake table with fake data...but on the other hand, I found the boolean filter condition easy to get wrong.  It's handy to know it works even with data like `(1, 1), (1, 2), (1, 3), (2, 1), (2, 2), (2, 3)` no matter where the marker is in that sequence.

File: nexus/src/db/pagination.rs [hunk 1]
```diff
@@ -61,3 +63,94 @@ where
...
+    BoxedQuery<T>:
+        query_methods::OrFilterDsl<Lt<C1, M1>, Output = BoxedQuery<T>>,
+{
+    let mut query = table.into_boxed().limit(pagparams.limit.get().into());
+    let marker = pagparams.marker.map(|m| m.clone());
+    match pagparams.direction {
+        dropshot::PaginationOrder::Ascending => {
+            if let Some((v1, v2)) = marker {
+                query = query.filter(c1.eq(v1.clone()).and(c2.gt(v2)));
+                query = query.or_filter(c1.gt(v1));
```
Comment (smklein) [inline nexus/src/db/pagination.rs:144]: Two functions provided by Diesel actually made these bounds a little easier for me to write, but admittedly different from your original implementation: 

https://docs.diesel.rs/master/diesel/prelude/trait.QueryDsl.html#method.or_filter
https://docs.diesel.rs/master/diesel/prelude/trait.QueryDsl.html#method.then_order_by

These are basically "optional versions" of filter and order, respectively, that can reduce the size of arguments passed.

They should be semantically equivalent to the "single-call, larger argument" version, but the trait bounds get a little less hairy.

**Review by david-crespo — APPROVED** [review]
nice!

**Comment by smklein** [comment]
> Nice! I'm curious how you figured out what the appropriate trait bounds are.

I mentioned this on our internal chat channel, but:

- Commenting out most of the implementation, and adding one operation at a time.
- Using mostly traits from "well-known" spots in Diesel, namely "QueryDSL" and "diesel's helper types".
- For the most part, ignoring suggestions from rustc.

> Is it worth adding a test? I'm not sure if it's worth generating a whole fake table with fake data...but on the other hand, I found the boolean filter condition easy to get wrong. It's handy to know it works even with data like `(1, 1), (1, 2), (1, 3), (2, 1), (2, 2), (2, 3)` no matter where the marker is in that sequence.

Sure, added tests for both the single and multi-column cases with a bespoke table for testing.

File: nexus/src/db/pagination.rs [hunk 2]
```diff
@@ -61,3 +63,358 @@ where
...
+                height: *height,
+            })
+            .collect();
+
+        diesel::insert_into(dsl::test_users)
+            .values(users)
+            .execute_async(pool.pool())
+            .await
+            .unwrap();
+    }
```
Comment (david-crespo) [inline nexus/src/db/pagination.rs:233]: nice. this is a good model for future tests

**Comment by smklein** [comment]
Closing this PR, as it was merged in with #512

**Pull request closed without merging**